File: src/StackQueue/Bst_Iterator.java
package StackQueue;

import java.util.Stack;

// https://www.interviewbit.com/problems/bst-iterator/
// next() and hasNext() should run in average O(1) time and uses O(h) memory, where h is the height of the tree.
public class Bst_Iterator {

    public class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;

        TreeNode(int x) {
            val = x;
        }
    }

    Stack<TreeNode> st = null;

    Bst_Iterator(TreeNode root) {
        st = new Stack<>();
        TreeNode curr = root;
        while (curr != null) {
            st.push(curr);
            curr = curr.left; // pushing only left node
        }
    }

    /** @return check next smallest number */
    public boolean hasNext() {
        return !st.empty();
    }

    /** @return the next smallest number */
    public int next() {
        if (st.empty()) {
            return -1;
        }
        TreeNode curr = st.pop();
        int val = curr.val;

        curr = curr.right; // cheking right node
        while (curr != null) {
            st.push(curr);
            curr = curr.left; // pusing all left node of right node
        }
        return val;
    }
}

File: src/leetcode/medium/BSTIterator.cpp
#include <iostream>

using namespace std;

struct TreeNode
{
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class BSTIterator
{
private:
    stack<TreeNode *> st;
    TreeNode *curr;

public:
    BSTIterator(TreeNode *root)
    {
        curr = root;
    }

    /**
     * @return the next smallest number
     */
    int next()
    {
        while (curr)
        {
            st.push(curr);
            curr = curr->left;
        }
        curr = st.top();
        st.pop();
        int ret = curr->val;
        curr = curr->right;
        return ret;
    }

    /**
     * @return whether we have a next smallest number
     */
    bool hasNext()
    {
        return curr || !st.empty();
    }
};

class BSTIterator2
{
private:
    vector<TreeNode *> inorder;

    vector<TreeNode *> inorderTraversal(TreeNode *root)
    {
        if (root == nullptr)
            return {};

        vector<TreeNode *> leftTraversal = inorderTraversal(root->left);
        vector<TreeNode *> rightTraversal = inorderTraversal(root->right);

        leftTraversal.push_back(root);
        leftTraversal.insert(leftTraversal.end(), rightTraversal.begin(), rightTraversal.end());
        return leftTraversal;
    }

public:
    BSTIterator2(TreeNode *root)
    {
        inorder = this->inorderTraversal(root);
        reverse(inorder.begin(), inorder.end());
    }

    int next()
    {
        TreeNode *ret = inorder.back();
        inorder.pop_back();
        return ret->val;
    }

    bool hasNext()
    {
        return !inorder.empty();
    }
};